internal/impl/nats/input_kv.go

```diff
@@ -0,0 +1,271 @@
+}
+
+func newKVReader(conf *service.ParsedConfig, mgr *service.Resources) (*kvReader, error) {
+	r := &kvReader{
+		log:     mgr.Logger(),
+		fs:      mgr.FS(),
+		shutSig: shutdown.NewSignaller(),
+	}
+
+	urlList, err := conf.FieldStringList("urls")
+	if err != nil {
+		return nil, err
+	}
+	r.urls = strings.Join(urlList, ",")
+
+	if r.bucket, err = conf.FieldString("bucket"); err != nil {
+		return nil, err
+	}
+
+	if r.ignoreDeletes, err = conf.FieldBool("ignore_deletes"); err != nil {
+		return nil, err
+	}
+
+	if r.includeHistory, err = conf.FieldBool("include_history"); err != nil {
+		return nil, err
+	}
+
+	if r.metaOnly, err = conf.FieldBool("meta_only"); err != nil {
+		return nil, err
+	}
+
+	if r.key, err = conf.FieldString("key"); err != nil {
+		return nil, err
+	}
+
+	tlsConf, tlsEnabled, err := conf.FieldTLSToggled("tls")
+	if err != nil {
+		return nil, err
+	}
+	if tlsEnabled {
+		r.tlsConf = tlsConf
+	}
+
+	if r.authConf, err = AuthFromParsedConfig(conf.Namespace("auth")); err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func (r *kvReader) Connect(ctx context.Context) error {
+	r.connMut.Lock()
+	defer r.connMut.Unlock()
+
+	if r.natsConn != nil {
+		return nil
+	}
+
+	var err error
+
+	defer func() {
+		if err != nil {
+			if r.watcher != nil {
+				_ = r.watcher.Stop()
+			}
+			if r.natsConn != nil {
+				r.natsConn.Close()
+			}
+		}
+	}()
+
+	var opts []nats.Option
+	if r.tlsConf != nil {
+		opts = append(opts, nats.Secure(r.tlsConf))
+	}
+	opts = append(opts, authConfToOptions(r.authConf, r.fs)...)
+	if r.natsConn, err = nats.Connect(r.urls, opts...); err != nil {
+		return err
+	}
+
+	js, err := r.natsConn.JetStream()
+	if err != nil {
+		return err
+	}
+
+	kv, err := js.KeyValue(r.bucket)
+	if err != nil {
+		return err
+	}
+
+	var watchOpts []nats.WatchOpt
+	if r.ignoreDeletes {
+		watchOpts = append(watchOpts, nats.IgnoreDeletes())
+	}
+	if r.includeHistory {
+		watchOpts = append(watchOpts, nats.IncludeHistory())
+	}
+	if r.metaOnly {
+		watchOpts = append(watchOpts, nats.MetaOnly())
+	}
+
+	r.watcher, err = kv.Watch(r.key, watchOpts...)
+	if err != nil {
+		return err
+	}
+
+	r.log.Infof("Watching NATS KV bucket: %s for key(s): %s", r.bucket, r.key)
+
+	return nil
+}
+
+func (r *kvReader) disconnect() {
+	r.connMut.Lock()
+	defer r.connMut.Unlock()
+
+	if r.watcher != nil {
+		_ = r.watcher.Stop()
+		r.watcher = nil
+	}
+	if r.natsConn != nil {
+		r.natsConn.Close()
+		r.natsConn = nil
+	}
+}
+
+func (r *kvReader) Read(ctx context.Context) (*service.Message, service.AckFunc, error) {
+	r.connMut.Lock()
+	watcher := r.watcher
+	r.connMut.Unlock()
+
+	if watcher == nil {
+		return nil, nil, service.ErrNotConnected
+	}
+
+	for {
+		var entry nats.KeyValueEntry
+		var open bool
+		select {
+		case entry, open = <-watcher.Updates():
+		case <-ctx.Done():
+			return nil, nil, ctx.Err()
+		}
+
+		if !open {
+			r.disconnect()
+			return nil, nil, service.ErrNotConnected
+		}
+
+		if entry == nil {
```

Comment:
I'm assuming this could happen if some internal timeout occurs? Just making sure we can't enter a busy loop here.